apps/builder/app/builder/features/ai/operations.tsx

```diff
@@ -0,0 +1,219 @@
+import { getMapValuesByKeysSet } from "~/shared/array-utils";
+import {
+  breakpointsStore,
+  dataSourcesStore,
+  instancesStore,
+  propsStore,
+  registeredComponentMetasStore,
+  styleSourceSelectionsStore,
+  stylesStore,
+} from "~/shared/nano-states";
+import { applyOperations } from "./apply-operations";
+import { useStore } from "@nanostores/react";
+import { restAiOperations } from "~/shared/router-utils";
+
+export const Operations = ({
+  rootInstanceId,
+}: {
+  rootInstanceId: Instance["id"];
+}) => {
+  const abort = useRef<AbortController>();
+  const [isLoading, setIsLoading] = useState(false);
+  const [error, setError] = useState<string>();
+
+  const jsx = useStore(useInstanceToJsx(rootInstanceId));
+
+  return (
+    <form
+      method="POST"
+      action=""
+      onSubmit={(event) => {
+        event.preventDefault();
+
+        if (isLoading) {
+          return;
+        }
+
+        const formData = new FormData(event.currentTarget);
+
+        const prompt = formData.get("prompt");
+        let data = formData.get("data");
+
+        if (typeof data === "string") {
+          try {
+            data = JSON.parse(data);
+          } catch {
+            /**/
+          }
+        }
+
+        if (typeof prompt !== "string" || typeof data !== "string") {
+          alert("Invalid data");
+          return;
+        }
+
+        const metas = registeredComponentMetasStore.get();
+        const exclude = ["Body", "Slot"];
+
+        const components = [...metas.keys()].filter(
+          (name) => !exclude.includes(name)
+        );
+
+        abort.current = new AbortController();
+
+        request<operations.Response>(
+          [
+            restAiOperations(),
+            {
+              method: "POST",
+              body: JSON.stringify({
+                prompt,
+                components,
+                jsx: data,
+              }),
+              signal: abort.current.signal,
+            },
+          ],
+          { retry: 2 }
+        )
+          .then((result) => {
+            if (result.success === false) {
+              setError(result.message);
+              return;
+            }
+            applyOperations(result.data);
+          })
+          .finally(() => {
+            abort.current = undefined;
+            setIsLoading(false);
+          });
+
+        setError(undefined);
+        setIsLoading(true);
+      }}
+    >
+      <Label>
+        Prompt
+        <InputField name="prompt" />
+      </Label>
+      <input type="hidden" name="data" value={JSON.stringify(jsx)} />
+      <Button
+        onClick={(event) => {
+          if (isLoading) {
+            event.preventDefault();
+            abort.current?.abort();
+          }
+        }}
+      >
+        {isLoading ? "stop" : "go"}
+      </Button>{" "}
+      {isLoading ? <SpinnerIcon /> : null}
+      {error ? <Text>{error}</Text> : null}
+    </form>
+  );
+};
+
+const useInstanceToJsx = (rootInstanceId: Instance["id"]) =>
```

Comment:
Github lost my comment.

This whole thing can be a top level computed with selectedInstanceSelector as input instead of passing it with prop.